### PR TITLE
Map service

### PR DESCRIPTION
**Pull request by truongtamthanh2004**
# ✅ Checklist Tự review
## 🧹 Code sạch & format chuẩn
- [x] Code sạch, dễ đọc, theo convention của team (naming, cấu trúc)
- [ ] Đã chạy `go fmt`, `goimports`, `golangci-lint` (nếu có)
- [x] Không còn đoạn code tạm, `fmt.Println`, `log.Println`, comment debug
- [x] Không còn biến, hàm, import không sử dụng
- [x] Hàm không quá dài, có thể tách nhỏ nếu vượt quá 50 dòng
- [x] Code không lặp lại, đã tái sử dụng hợp lý

## 🧠 Logic & Xử lý lỗi
- [x] Logic đúng yêu cầu, không dư thừa, rõ ràng
- [x] Xử lý đầy đủ các trường hợp `nil`, `error`, `invalid input`
- [x] Mỗi thao tác có thể lỗi (I/O, DB, parse...) đều được check `err`
- [x] Không để `err` bị silent — luôn log hoặc trả về (tùy ngữ cảnh)
- [x] Sử dụng `context` đúng cách (nếu dùng), tránh leak
- [x] Nếu là `goroutine`, có quản lý lifecycle và tránh leak

## 🌐 HTTP Handler (Gin)
- [x] Sử dụng đúng HTTP method (GET, POST, PUT, DELETE…)
- [x] Validate request (param, query, body) đầy đủ, rõ ràng
- [x] Sử dụng binding (`ShouldBindJSON`, `ShouldBindUri`, v.v.) đúng cách
- [x] Xử lý trả về `status code` phù hợp (`200`, `201`, `400`, `404`, `500`…)
- [x] Response rõ ràng, có `message`, `data`, `error` (nếu cần)
- [x] Không leak thông tin nhạy cảm trong response lỗi

## 🗃️ Struct & Layer
- [x] Struct có annotation JSON đầy đủ (`json:"field_name"`)
- [x] Sử dụng DTO/input/output rõ ràng, không lẫn lộn model DB và response
- [x] Đã tách rõ các layer: handler → service → repository (nếu theo clean code)
- [x] Middleware viết rõ ràng, dễ đọc, không làm chậm hệ thống

## 🧪 Unit Test & Mock
- [x] Có test cho handler/service chính (ít nhất happy case và edge case)
- [x] Nếu dùng DB/service ngoài, có mock hoặc test isolation
- [x] Có test validate đầu vào (request, param, query)
- [x] Không dùng test hardcode giá trị không ổn định (vd: time.Now() mà không mock)
- [x] Tên test rõ ràng: `TestFunction_Scenario_Expected`

## 🔐 Security & Config
- [x] Không hardcode secret, token, config trong code
- [x] Không để lộ thông tin nhạy cảm qua log hoặc panic
- [x] Validate và sanitize đầu vào người dùng
- [x] Nếu có auth, kiểm tra middleware hoạt động đúng

## 🤖 Sử dụng GitHub Copilot
- [x] Đã kiểm tra kỹ mọi đoạn code được gợi ý từ GitHub Copilot
- [x] Không để lại phần comment hoặc TODO do Copilot sinh ra
- [x] Không commit đoạn code mà chưa hiểu rõ logic hoặc mục đích
- [x] Đảm bảo mọi function/code block từ Copilot phù hợp với context dự án
- [x] Tối ưu lại code được gợi ý nếu cần — tránh lạm dụng hoặc copy nguyên mẫu

## Related Tickets
- ticket redmine link

## WHAT (optional)
- Change number items `completed/total` in admin page.

## HOW
- I edit js file, inject not_vary_normal items in calculate function.

## WHY (optional)
- Because in previous version - number just depends on `normal` items. But in new version, we have `state` and `confirm_state` depends on both `normal` + `not_normal` items.

### CODE REVIEW

File: services/map-service/internal/service/geocode_service.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+func GeocodeAddress(address string) (GeoCodingResult, error) {
+	apiKey := os.Getenv("GOOGLE_MAPS_API_KEY")
+	baseURL := "https://maps.googleapis.com/maps/api/geocode/json"
+
+	u := fmt.Sprintf("%s?address=%s&key=%s", baseURL, url.QueryEscape(address), apiKey)
+	resp, err := http.Get(u)
+	if err != nil {
+		return GeoCodingResult{}, err
+	}
```
Comment (chienpv-3590) [inline services/map-service/internal/service/geocode_service.go:24]: Hiện tại chỉ kiểm tra lỗi khi gọi GET, nhưng chưa kiểm tra status code của response (nên thêm kiểm tra resp.StatusCode != http.StatusOK)

File: services/map-service/internal/service/geocode_service.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	"os"
+)
+
+type GeoCodingResult struct {
+	Lat float64
+	Lng float64
+}
+
+func GeocodeAddress(address string) (GeoCodingResult, error) {
+	apiKey := os.Getenv("GOOGLE_MAPS_API_KEY")
```
Comment (chienpv-3590) [inline services/map-service/internal/service/geocode_service.go:17]: Nên kiểm tra trường hợp biến môi trường API key không tồn tại hoặc rỗng, trả về lỗi rõ ràng.

File: services/map-service/internal/service/venue_service.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	cacheKey := "venues:all"
+
+	if data, err := config.Get(ctx, cacheKey); err == nil {
+		var venues []dto.Venue
+		if json.Unmarshal([]byte(data), &venues) == nil {
+			return venues, nil
+		}
+	}
+
+	resp, err := http.Get(fmt.Sprintf("%s/api/v1/venues", s.BaseURL))
```
Comment (chienpv-3590) [inline services/map-service/internal/service/venue_service.go:41]: Nên sử dụng http.Client với timeout, tránh trường hợp request treo lâu.

File: services/map-service/internal/usecase/map_usecase.go [hunk 4]
```diff
@@ -1 +1,60 @@
...
+	for i, v := range venues {
+		coord, err := service.GeocodeAddress(fmt.Sprintf("%s, %s", v.Address, v.City))
+		if err == nil {
+			venues[i].Latitude = coord.Lat
+			venues[i].Longitude = coord.Lng
+		}
+	} 
+
+	jsonStr, err := toJSON(venues)
+	if err != nil {
```
Comment (chienpv-3590) [inline services/map-service/internal/usecase/map_usecase.go:43]: Nếu geocode lỗi, hiện tại chỉ bỏ qua, không log hoặc đánh dấu venue nào bị thiếu vị trí. Nên log lỗi hoặc có cơ chế fallback

File: services/map-service/test/map_usecase_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package test
```
Comment (chienpv-3590) [inline services/map-service/test/map_usecase_test.go:1]: Package test nên đặt là usecase_test để rõ ràng hơn về vị trí và mục đích (hoặc đặt cùng package với usecase để test private)

File: services/map-service/test/map_usecase_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	stub := &StubVenueService{
+		FetchVenuesFunc: func() ([]dto.Venue, error) {
+			return nil, errors.New("fetch failed")
+		},
+	}
+	uc := usecase.NewMapUsecase(stub)
+
+	// Act
+	venues, err := uc.GetVenues()
+
```
Comment (chienpv-3590) [inline services/map-service/test/map_usecase_test.go:51]: Nên bổ sung test trả về []dto.Venue{} để kiểm tra hàm xử lý mảng rỗng.

**Comment by truongtamthanh2004** [comment]
ready
